File: backend/src/domain/entities/Product.ts
import type { NormalizedUnit } from "../value-objects/NormalizedUnit";
import type { Price } from "../value-objects/Price";
import type { ProductCategory } from "../value-objects/ProductCategory";
import type { Supermarket } from "../value-objects/Supermarket";
import type { TaxType } from "../value-objects/TaxType";

export interface Product {
	readonly id: string;
	readonly sku: string;
	readonly name: string;
	readonly supermarket: Supermarket;
	readonly category: ProductCategory;
	readonly price: Price;
	readonly pricePerUnit: Price;
	readonly unit: NormalizedUnit;
	readonly taxType: TaxType;
	readonly scrapedAt: Date;
	readonly image?: string;
	readonly url?: string;
}

File: backend-dotnet/src/api/Domain/Entities/Product.cs
using SupermarketPlanner.Api.Domain.Models;
using SupermarketPlanner.Api.Domain.ValueObjects;

namespace SupermarketPlanner.Api.Domain.Entities;

public sealed record Product
{
    private Product(
        string id,
        string sku,
        string name,
        SupermarketName supermarket,
        ProductCategory category,
        Price price,
        Price pricePerUnit,
        NormalizedUnit unit,
        TaxType tax,
        DateTimeOffset scrapedAt,
        string? image,
        string? url)
    {
        Id = id;
        Sku = sku;
        Name = name;
        Supermarket = supermarket;
        Category = category;
        Price = price;
        PricePerUnit = pricePerUnit;
        Unit = unit;
        Tax = tax;
        ScrapedAt = scrapedAt;
        Image = image;
        Url = url;
    }

    public string Id { get; }
    public string Sku { get; }
    public string Name { get; }
    public SupermarketName Supermarket { get; }
    public ProductCategory Category { get; }
    public Price Price { get; }
    public Price PricePerUnit { get; }
    public NormalizedUnit Unit { get; }
    public TaxType Tax { get; }
    public DateTimeOffset ScrapedAt { get; }
    public string? Image { get; }
    public string? Url { get; }

    public static Product Scraped(
        string id,
        string sku,
        string name,
        SupermarketName supermarket,
        ProductCategory category,
        Price price,
        Price pricePerUnit,
        NormalizedUnit unit,
        TaxType tax,
        DateTimeOffset scrapedAt,
        string? image,
        string? url) =>
        new(id, sku, name, supermarket, category, price, pricePerUnit, unit, tax, scrapedAt, image, url);
}
